### PR TITLE
[Issue #18] Add support for VApp Resource

### PR DESCRIPTION
**Pull request by maheshd2**
Added support for VApp Resource
-> Cloning of VApp is supported
-> Delete of VApp is supported

### CODE REVIEW

File: vsphere/resource_vsphere_vapp.go [hunk 1]
```diff
@@ -0,0 +1,723 @@
...
+			},
+			"folder": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"parent_vapp": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"entity": &schema.Schema{
```
Comment (chandidas) [inline vsphere/resource_vsphere_vapp.go]: 'entity' can be leftout for now.

File: vsphere/resource_vsphere_vapp.go [hunk 2]
```diff
@@ -0,0 +1,723 @@
...
+}
+
+func resourceVSphereVAppCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*govmomi.Client)
+
+	log.Printf("[DEBUG] resourceVSphereVAppCreate :: client : %#v", client)
+	log.Printf("[DEBUG] resourceVSphereVAppCreate :: ResourceData d: %#v", d)
+
+	// Construct vAPP Object with required Attributes
+	vapp := vApp{
```
Comment (chandidas) [inline vsphere/resource_vsphere_vapp.go]: Create a NewVApp for this.

File: vsphere/resource_vsphere_vapp.go [hunk 3]
```diff
@@ -0,0 +1,723 @@
...
+		d:    d,
+		c:    client,
+		name: d.Get("name").(string),
+	}
+
+	err := vapp.populateOptionalVAppAttributes(d)
+	if err != nil {
+		log.Printf("[ERROR] %s", err)
```
Comment (chandidas) [inline vsphere/resource_vsphere_vapp.go]: write more elaborate error message.

File: vsphere/resource_vsphere_vapp.go [hunk 4]
```diff
@@ -0,0 +1,723 @@
...
+		return err
+	}
+
+	err = vapp.linkEntities()
+	if err != nil {
+		log.Printf("[ERROR] %s", err)
+		return err
+	}
+	d.SetId(vapp.name)
```
Comment (chandidas) [inline vsphere/resource_vsphere_vapp.go]: Instead of name set complete path.

File: vsphere/resource_vsphere_vapp.go [hunk 5]
```diff
@@ -0,0 +1,723 @@
...
+	log.Printf("[DEBUG] resourceVSphereVAppRead:: d : %#v", d)
+
+	d.Set("name", d.Get("name"))
+
+	log.Printf("[DEBUG] resourceVSphereVAppRead:: d : %#v", d)
+
+	return nil
+}
+func resourceVSphereVAppUpdate(d *schema.ResourceData, meta interface{}) error {
+	return nil
```
Comment (chandidas) [inline vsphere/resource_vsphere_vapp.go]: Add a debug statement like "not supported yet"

File: vsphere/resource_vsphere_vapp.go [hunk 6]
```diff
@@ -0,0 +1,723 @@
...
+}
+
+func resourceVSphereVAppDelete(d *schema.ResourceData, meta interface{}) error {
+
+	client := meta.(*govmomi.Client)
+
+	log.Printf("[DEBUG] resourceVSphereVAppDelete:: ResourceData d: %#v", d)
+
+	dc, err := getDatacenter(client, d.Get("datacenter").(string))
```
Comment (chandidas) [inline vsphere/resource_vsphere_vapp.go]: Finder object creation can be done inside getCreatedVApp.

File: vsphere/resource_vsphere_vapp.go [hunk 7]
```diff
@@ -0,0 +1,723 @@
...
+	finder = finder.SetDatacenter(dc)
+
+	vapp, err := getCreatedVApp(d, finder)
+	if err != nil {
+		return err
+	}
+
+	err = powerOffVApp(vapp)
+	if err != nil {
+		return err
```
Comment (chandidas) [inline vsphere/resource_vsphere_vapp.go]: Add ERROR log message too.

File: vsphere/resource_vsphere_vapp.go [hunk 8]
```diff
@@ -0,0 +1,723 @@
...
+		Update: resourceVSphereVAppUpdate,
+		Delete: resourceVSphereVAppDelete,
+
+		Schema: map[string]*schema.Schema{
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"description": &schema.Schema{
+				Type:     schema.TypeString,
```
Comment (maheshd2) [inline vsphere/resource_vsphere_vapp.go]: Added Default value for description
Default:  "Created by Terraform",

File: vsphere/resource_vsphere_vapp.go (hunk 2 above)
Comment (maheshd2) [inline vsphere/resource_vsphere_vapp.go]: Created NewVApp function

File: vsphere/resource_vsphere_vapp.go (hunk 3 above)
Comment (maheshd2) [inline vsphere/resource_vsphere_vapp.go]: Updated error messages

File: vsphere/resource_vsphere_vapp.go (hunk 4 above)
Comment (maheshd2) [inline vsphere/resource_vsphere_vapp.go]: Added New function to calculate the VApp path and set as Id

File: vsphere/resource_vsphere_vapp.go (hunk 5 above)
Comment (maheshd2) [inline vsphere/resource_vsphere_vapp.go]: Added meaningful debug statement

File: vsphere/resource_vsphere_vapp.go (hunk 6 above)
Comment (maheshd2) [inline vsphere/resource_vsphere_vapp.go]: This function is used in both delete and create operation. Hence only common things are placed under getCreatedVApp( ) function.

File: vsphere/resource_vsphere_vapp.go (hunk 7 above)
Comment (maheshd2) [inline vsphere/resource_vsphere_vapp.go]: Added ERROR messages to logs

**Pull request merged**